### ISSUE TITLE
reward not reflecting progress

### ISSUE THREAD

**Issue by ugokanain** [openai/gym#121]
I was playing with FrozenLake-v0, and though it has a simple logic, it has a hard time solving it. 

I found that the reward doesn't reflect the progress of the solution. For example, the reward is `rew = float(newletter == b'G')`, but wouldn't it be better something like 
`rew = float(newstate)`
`if newletter == 'H':`
`         new = -1`

Makes me think that most of the other environments might not reward progress, and confusses the neural net

**Comment by jietang** [comment]
This is by design - one interesting problem in RL is temporal credit assignment: learning what actions are responsible for earning rewards, even if the reward occurs much later.

You are correct that a smoothing the reward function to rewards progress would be easier to solve - it may be worth it to make sure your agent solves this easier problem before attempting the original FrozenLake.

**Comment by ugokanain** [comment]
Neither way it is really learning. I would had expected that if it truly learned, it should be able to solve the puzzle in about 10 steps (min 6 steps). However, after solving it several times, it still takes anywhere from 10 to 1000 steps to reach the goal. Seems to solve it randomly

By the way, I am not using the slippery version.

Maybe my code, but it is rather generic
`import gym
env = gym.make('FrozenLake-v0')
observation = env.reset()
for i_episode in range(10):
    for t in range(1000):
        env.render()
        action = env.action_space.sample()
        observation, reward, done, info = env.step(action)
	if done:
	    print observation, reward, done
            print("Episode finished after {} timesteps".format(t+1))
	    observation = env.reset()    
`

**Comment by jietang** [comment]
It looks like you are sampling a random action each time step - take a look at examples/agents/cem.py for an example of a learning agent (it implements the cross-entropy method, a black box optimization technique).

Closing this issue since there doesn't appear to be anything to fix.
